firmware/spade/src/shared/sprig_engine/module_native.c

```diff
@@ -656,12 +655,152 @@ JERRYXX_FUN(native_text_clear_fn) {
   dbg("module_native::native_text_clear_fn");
   text_clear(); return jerry_create_undefined(); }
 
+JERRYXX_FUN(set_value_fn) {
+  dbg("module_native::set_value_fn");
+
+  if (state->game_id == NULL) {
+      printf("Game ID is not set\n");
+      return jerry_create_undefined();
+  }
+
+  char *key = temp_str_mem();
+  jerry_size_t nbytes = jerry_string_to_char_buffer(
+    JERRYXX_GET_ARG(0),
+    (jerry_char_t *)key,
+    sizeof(state->temp_str_mem) - 1
+  );
+  key[nbytes] = '\0'; 
+
+  FIL fil;
+  int namebytes = strlen(state->game_id) + nbytes + 10;
+  char *filename = (char*)malloc(namebytes * sizeof(char));
+  if (filename == NULL) {
+    printf("malloc failed\n");
+    return jerry_create_undefined();
+  }
+  snprintf(filename, namebytes, "sprig/kv/%s/%s", state->game_id, key);
+  FRESULT fr = f_open(&fil, filename, FA_CREATE_ALWAYS | FA_WRITE);
+
+  char *data = temp_str_mem();
+  nbytes = jerry_string_to_char_buffer(
+    JERRYXX_GET_ARG(1),
+    (jerry_char_t *)data,
+    sizeof(state->temp_str_mem) - 1
+  );
+  data[nbytes] = '\0'; 
+
+  printf("writing %s to %s", data, filename);
+
+  int bw;
+  if (FR_OK != fr && FR_EXIST != fr)
+      panic("f_open(%s) error: %s (%d)\n", filename, FRESULT_str(fr), fr);
+
+  fr = f_close(&fil);
+  if (FR_OK != fr) {
+      printf("f_close error: %s (%d)\n", FRESULT_str(fr), fr);
+  }
+
+  free(filename);
+
+  return jerry_create_undefined();
+}
+
+JERRYXX_FUN(get_value_fn) {
+  dbg("module_native::get_value_fn");
+
+  if (state->game_id == NULL) {
+    printf("Game ID is not set\n");
+    return jerry_create_undefined();
+  }
+
+  char *key = temp_str_mem();
+  jerry_size_t nbytes = jerry_string_to_char_buffer(
+    JERRYXX_GET_ARG(0),
+    (jerry_char_t *)key,
+    sizeof(state->temp_str_mem) - 1
+  );
+  key[nbytes] = '\0'; 
+
+  FIL fil;
+  int namebytes = strlen(state->game_id) + nbytes + 10;
+  char *filename = (char*)malloc(namebytes * sizeof(char));
+  if (filename == NULL) {
+    printf("malloc failed\n");
+    return jerry_create_undefined();
+  }
+  snprintf(filename, namebytes, "sprig/kv/%s/%s", state->game_id, key);
+  FRESULT fr = f_open(&fil, filename, FA_READ);
+
+  if (FR_OK != fr) {
+    printf("f_open(%s) error: %s (%d)\n", filename, FRESULT_str(fr), fr);
+    return jerry_create_undefined();
```

Comment:
If this ends up happening, we lead the filename that was allocated successfully above.

This is one reason why a common pattern in reliable C is to use the following (yes, somewhat terrible, but trustworthy if done reliably):

(pseudocode)

```
....
int retCode = ERROR;
char *buffer = malloc(1);
if (buffer == null) {
  goto error;
}

char *buffer2 = malloc(1);
if (buffer2 == null) {
  goto error;
}
retCode = SUCCESS;

error:
CHKFREE(buffer);
CHKFREE(buffer2);
return retCode;
```

Reply:
where CHKFREE is a preprocessor define that checks for null, and frees if non-null

Reply:
thanks! not sure if I implemented this correctly, but I think i've stopped the leak - see commit https://github.com/hackclub/sprig/pull/1526/commits/5a358adfbb87e0bfca2b6db310adaab269864ec8